### PR TITLE
Jfrog Artifactory service endpoint

### PR DESCRIPTION
**Pull request by billbirchatcoles**
This PR adds a new resource for service endpoint for JFrog Artifactory for either username/password or api-tokens.

## All Submissions:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] I have updated the documentation accordingly.
* [x] I have added tests to cover my changes.
* [x] All new and existing tests passed.
* [x] My code follows the code style of this project.
* [x] I ran lint checks locally prior to submission.
* [x] Have you checked to ensure there aren't other open PRs for the same update/change?

## What about the current behavior has changed?

None

Issue Number: #194 

## Does this introduce a change to `go.mod`, `go.sum` or `vendor/`?

- [ ] Yes
- [x] No

## Does this introduce a breaking change?

- [ ] Yes
- [X] No

## Any relevant logs, error output, etc?

(If it’s long, please paste to https://ghostbin.com/ and insert the link here.)

## Other information

This provider requires the JFrog Artifactory Extension to be installed in the Organisation

>   Extension details
>   Integrate your JFrog Artifactory with Visual Studio Team Services.
>   Publisher
>   JFrog
>   Installed version
>   1.11.0 (Latest)
>   Last updated
>   Nov 17, 2020 at 3:19 AM GMT+11
>

### CODE REVIEW

**Comment by billbirchatcoles** [comment]
Tests are failing since the JFrog extension is not installed.

This provider requires the JFrog Artifactory Extension to be installed in the Organisation.

> Extension details
> Integrate your JFrog Artifactory with Visual Studio Team Services.
> Publisher
> JFrog
> Installed version
> 1.11.0 (Latest)
> Last updated
> Nov 17, 2020 at 3:19 AM GMT+11

**Comment by billbirchatcoles** [comment]
Hi @xuzhang3 ,

This provider needs the JFrog Artifactory extension installed in the Org in order that the tests can run. Is that something I can do? Or is this something only you can control?

Regards,
Bill

**Comment by xuzhang3** [comment]
Hi @billbirchatcoles I will install the extension for the test ORG and rerun the tests.

File: azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_artifactory.go [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+	r.Schema["authentication_token"] = &schema.Schema{
+		Type:         schema.TypeList,
+		Optional:     true,
+		MinItems:     1,
+		MaxItems:     1,
+		Elem:         at,
+		ExactlyOneOf: []string{"authentication_password", "authentication_token"},
+	}
+
+	r.Schema["authentication_password"] = &schema.Schema{
```
Comment (xuzhang3) [inline azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_artifactory.go]: I would suggest rename `authentication_password` to `authentication_basic`

File: azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_artifactory.go [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+			}
+			return validation.IsURLWithHTTPorHTTPS(url, key)
+		},
+		Description: "Url for the Artifactory Server",
+	}
+
+	patHashKey, patHashSchema := tfhelper.GenerateSecreteMemoSchema("access_token")
+	at := &schema.Resource{
+		Schema: map[string]*schema.Schema{
+			"access_token": {
```
Comment (xuzhang3) [inline azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_artifactory.go]: I would suggest rename `access_token` to `token`

File: website/docs/r/serviceendpoint_artifactory.html.markdown [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+  project_id            = azuredevops_project.project.id
+  service_endpoint_name = "artifactory-mainframe"
+  description           = "My Service Connection"
+  url                   = "https://artifactory.mainframe.com"
+  authentication_token {
+      access_token      = "REDACTEDREDACTED"
+  }
+}
+```
+Alternativley a username and password may be used.
```
Comment (xuzhang3) [inline website/docs/r/serviceendpoint_artifactory.html.markdown]:
```suggestion
Alternatively a username and password may be used.
```

File: azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_artifactory.go [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+				newHash, hashKey := tfhelper.HelpFlattenSecretNested(d, "authentication_token", authList, "access_token")
+				auth[hashKey] = newHash
+			}
+		}
+		if serviceEndpoint.Authorization != nil && serviceEndpoint.Authorization.Parameters != nil {
+			auth["access_token"] = (*serviceEndpoint.Authorization.Parameters)["apitoken"]
+		}
+		d.Set("authentication_token", []interface{}{auth})
+	}
+	if strings.EqualFold(*serviceEndpoint.Authorization.Scheme, "UsernamePassword") {
```
Comment (xuzhang3) [inline azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_artifactory.go]: I would suggest change `if` to `else if`

File: azuredevops/internal/acceptancetests/resource_serviceendpoint_artifactory_test.go [hunk 5]
```diff
@@ -0,0 +1,373 @@
...
+					resource.TestCheckResourceAttrSet(tfSvcEpNode, "project_id"),
+					resource.TestCheckResourceAttr(tfSvcEpNode, "authentication_password.#", "1"),
+					resource.TestCheckResourceAttr(tfSvcEpNode, "service_endpoint_name", serviceEndpointName),
+				),
+			},
+		},
+	})
+}
+
+func TestAccServiceEndpointArtifactory_complete(t *testing.T) {
```
Comment (xuzhang3) [inline azuredevops/internal/acceptancetests/resource_serviceendpoint_artifactory_test.go]:
```suggestion
func TestAccServiceEndpointArtifactory_complete_token(t *testing.T) {
```

**Review by xuzhang3 — CHANGES_REQUESTED** [review]
Few changes requested but otherwise looks good

File: azuredevops/internal/acceptancetests/resource_serviceendpoint_artifactory_test.go (hunk 5 above)
Comment (birchb1024) [reply azuredevops/internal/acceptancetests/resource_serviceendpoint_artifactory_test.go]: Done

File: azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_artifactory.go (hunk 4 above)
Comment (birchb1024) [reply azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_artifactory.go]: Done

File: website/docs/r/serviceendpoint_artifactory.html.markdown (hunk 3 above)
Comment (birchb1024) [reply website/docs/r/serviceendpoint_artifactory.html.markdown]: Done

File: azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_artifactory.go (hunk 2 above)
Comment (birchb1024) [reply azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_artifactory.go]: Done

File: azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_artifactory.go (hunk 1 above)
Comment (birchb1024) [reply azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_artifactory.go]: Done

**Comment by xuzhang3** [comment]
Hi @billbirchatcoles Can you help handle the branch conflict?

**Comment by birchb1024** [comment]
Merged in latest master (3bd4c54). Conflict resolved. Retested.

**Comment by xuzhang3** [comment]
Hi @birchb1024  I don't have the push permission of you branch,   can you help resolve the conflict?

**Comment by birchb1024** [comment]
Merged in latest master (76e37e7). Conflict resolved. Retested.

**Comment by xuzhang3** [comment]
LGTM

**Pull request merged**